make cover: to many open files

I'm guessing this is because log files are not closed in some modules.

```
$ make cover pkgs=./modules/consensus
rm -rf release doc/whitepaper.aux doc/whitepaper.log doc/whitepaper.pdf
--- FAIL: TestIntegrationBroadcastRelayHeader (0.02s)
	synchronize_test.go:1100: open /var/folders/7q/hm_qblpd4h303j504dfg2zk40000gn/T/SiaTesting/consensus/TestIntegrationBroadcastRelayHeader2/wallet/wallet.log: too many open files
--- FAIL: TestIntegrationRelaySynchronize (0.00s)
	synchronize_test.go:1155: error opening consensus database: open /var/folders/7q/hm_qblpd4h303j504dfg2zk40000gn/T/SiaTesting/consensus/TestRelaySynchronize1/consensus/consensus.db: too many open files
--- FAIL: TestThreadedReceiveBlocksStalls (0.00s)
	synchronize_test.go:1286: open /var/folders/7q/hm_qblpd4h303j504dfg2zk40000gn/T/SiaTesting/consensus/TestThreadedReceiveBlocksStalls/gateway/gateway.log: too many open files
--- FAIL: TestIntegrationSendBlocksStalls (0.00s)
	synchronize_test.go:1403: open /var/folders/7q/hm_qblpd4h303j504dfg2zk40000gn/T/SiaTesting/consensus/TestThreadedReceiveBlocksTimesout - local/gateway/gateway.log: too many open files
--- FAIL: TestTryValidTransactionSet (0.00s)
	validtransaction_test.go:17: open /var/folders/7q/hm_qblpd4h303j504dfg2zk40000gn/T/SiaTesting/consensus/TestValidTransaction/gateway/gateway.log: too many open files
--- FAIL: TestTryInvalidTransactionSet (0.00s)
	validtransaction_test.go:48: open /var/folders/7q/hm_qblpd4h303j504dfg2zk40000gn/T/SiaTesting/consensus/TestValidTransaction/gateway/gateway.log: too many open files
```

--- a/api/server.go
+++ b/api/server.go
@@ -120,7 +120,11 @@ func (srv *Server) Close() error {
 			errStrs = append(errStrs, fmt.Sprintf("explorer err: %v", err))
 		}
 	}
-	// TODO: close miner
+	if srv.miner != nil {
+		if err := srv.miner.Close(); err != nil {
+			errStrs = append(errStrs, fmt.Sprintf("miner err: %v", err))
+		}
+	}
 	if srv.wallet != nil {
 		// TODO: close wallet and lock the wallet in the wallet's Close method.
 		if srv.wallet.Unlocked() {

--- a/api/server.go
+++ b/api/server.go
@@ -76,14 +76,20 @@ func (srv *Server) Serve() error {
 	}
 
 	// safely close each module
+	if srv.host != nil {
+		srv.host.Close()
+	}
+	if srv.explorer != nil {
+		srv.explorer.Close()
+	}
+	if srv.wallet != nil {
+		srv.wallet.Lock()
+	}
 	if srv.cs != nil {
 		srv.cs.Close()
 	}
 	if srv.gateway != nil {
 		srv.gateway.Close()
-	}
-	if srv.wallet != nil {
-		srv.wallet.Lock()
 	}
 
 	return nil

--- a/api/server.go
+++ b/api/server.go
@@ -1,6 +1,7 @@
 package api
 
 import (
+	"errors"
 	"fmt"
 	"net"
 	"net/http"
@@ -67,31 +68,52 @@ func (srv *Server) Serve() error {
 		srv.listener.Close()
 	}()
 
+	var errStrs []string
+
 	// The server will run until an error is encountered or the listener is
 	// closed, via either the Close method or the signal handling above.
 	// Closing the listener will result in the benign error handled below.
 	err := srv.apiServer.Serve(srv.listener)
 	if err != nil && !strings.HasSuffix(err.Error(), "use of closed network connection") {
-		return err
+		errStrs = append(errStrs, fmt.Sprintf("serve err: %v", err))
 	}
 
 	// safely close each module
 	if srv.host != nil {
-		srv.host.Close()
+		if err := srv.host.Close(); err != nil {
+			errStrs = append(errStrs, fmt.Sprintf("host err: %v", err))
+		}
 	}
+	// TODO: close renter (which should close hostdb as well)
 	if srv.explorer != nil {
-		srv.explorer.Close()
+		if err := srv.explorer.Close(); err != nil {
+			errStrs = append(errStrs, fmt.Sprintf("explorer err: %v", err))
+		}
 	}
+	// TODO: close miner
 	if srv.wallet != nil {
-		srv.wallet.Lock()
+		// TODO: close wallet and lock the wallet in the wallet's Close method.
+		if srv.wallet.Unlocked() {
+			if err := srv.wallet.Lock(); err != nil {
+				errStrs = append(errStrs, fmt.Sprintf("wallet err: %v", err))
+			}
+		}
 	}
+	// TODO: close transaction pool
 	if srv.cs != nil {
-		srv.cs.Close()
+		if err := srv.cs.Close(); err != nil {
+			errStrs = append(errStrs, fmt.Sprintf("consensus err: %v", err))
+		}
 	}
 	if srv.gateway != nil {
-		srv.gateway.Close()
+		if err := srv.gateway.Close(); err != nil {
+			errStrs = append(errStrs, fmt.Sprintf("gateway err: %v", err))
+		}
 	}
 
+	if len(errStrs) > 0 {
+		return errors.New(strings.Join(errStrs, "\n"))
+	}
 	return nil
 }
 

--- a/api/server.go
+++ b/api/server.go
@@ -69,13 +69,19 @@ func (srv *Server) Serve() error {
 	defer srv.wg.Done()
 
 	// stop the server if a kill signal is caught
-	sigChan := make(chan os.Signal)
+	sigChan := make(chan os.Signal, 1)
 	signal.Notify(sigChan, os.Interrupt, os.Kill)
-	defer signal.Reset(os.Interrupt, os.Kill)
+	defer signal.Stop(sigChan)
+	stop := make(chan struct{})
+	defer close(stop)
 	go func() {
-		<-sigChan
-		fmt.Println("\rCaught stop signal, quitting...")
-		srv.listener.Close()
+		select {
+		case <-sigChan:
+			fmt.Println("\rCaught stop signal, quitting...")
+			srv.listener.Close()
+		case <-stop:
+			// Don't leave a dangling goroutine.
+		}
 	}()
 
 	// The server will run until an error is encountered or the listener is

--- a/api/server.go
+++ b/api/server.go
@@ -1,7 +1,6 @@
 package api
 
 import (
-	"errors"
 	"fmt"
 	"net"
 	"net/http"
@@ -10,6 +9,7 @@ import (
 	"strings"
 	"sync"
 
+	"github.com/NebulousLabs/Sia/build"
 	"github.com/NebulousLabs/Sia/modules"
 )
 
@@ -96,11 +96,11 @@ func (srv *Server) Serve() error {
 
 // Close closes the Server's listener, causing the HTTP server to shut down.
 func (srv *Server) Close() error {
-	var errStrs []string
+	var errs []error
 
 	// Close the listener, which will cause Server.Serve() to return.
 	if err := srv.listener.Close(); err != nil {
-		errStrs = append(errStrs, fmt.Sprintf("listener err: %v", err))
+		errs = append(errs, fmt.Errorf("listener.Close failed: %v", err))
 	}
 
 	// Wait for Server.Serve() to exit. We wait so that it's guaranteed that the
@@ -111,42 +111,39 @@ func (srv *Server) Close() error {
 	// Safely close each module.
 	if srv.host != nil {
 		if err := srv.host.Close(); err != nil {
-			errStrs = append(errStrs, fmt.Sprintf("host err: %v", err))
+			errs = append(errs, fmt.Errorf("host.Close failed: %v", err))
 		}
 	}
 	// TODO: close renter (which should close hostdb as well)
 	if srv.explorer != nil {
 		if err := srv.explorer.Close(); err != nil {
-			errStrs = append(errStrs, fmt.Sprintf("explorer err: %v", err))
+			errs = append(errs, fmt.Errorf("explorer.Close failed: %v", err))
 		}
 	}
 	if srv.miner != nil {
 		if err := srv.miner.Close(); err != nil {
-			errStrs = append(errStrs, fmt.Sprintf("miner err: %v", err))
+			errs = append(errs, fmt.Errorf("miner.Close failed: %v", err))
 		}
 	}
 	if srv.wallet != nil {
 		// TODO: close wallet and lock the wallet in the wallet's Close method.
 		if srv.wallet.Unlocked() {
 			if err := srv.wallet.Lock(); err != nil {
-				errStrs = append(errStrs, fmt.Sprintf("wallet err: %v", err))
+				errs = append(errs, fmt.Errorf("wallet.Lock failed: %v", err))
 			}
 		}
 	}
 	// TODO: close transaction pool
 	if srv.cs != nil {
 		if err := srv.cs.Close(); err != nil {
-			errStrs = append(errStrs, fmt.Sprintf("consensus err: %v", err))
+			errs = append(errs, fmt.Errorf("consensusset.Close failed: %v", err))
 		}
 	}
 	if srv.gateway != nil {
 		if err := srv.gateway.Close(); err != nil {
-			errStrs = append(errStrs, fmt.Sprintf("gateway err: %v", err))
+			errs = append(errs, fmt.Errorf("gateway.Close failed: %v", err))
 		}
 	}
 
-	if len(errStrs) > 0 {
-		return errors.New(strings.Join(errStrs, "\n"))
-	}
-	return nil
+	return build.JoinErrors(errs, "\n")
 }

--- a/api/server.go
+++ b/api/server.go
@@ -8,6 +8,7 @@ import (
 	"os"
 	"os/signal"
 	"strings"
+	"sync"
 
 	"github.com/NebulousLabs/Sia/modules"
 )
@@ -27,6 +28,11 @@ type Server struct {
 	apiServer         *http.Server
 	listener          net.Listener
 	requiredUserAgent string
+
+	// wg is used to block Close() from returning until Serve() has finished. A
+	// WaitGroup is used instead of a chan struct{} so that Close() can be called
+	// without necessarily calling Serve() first.
+	wg sync.WaitGroup
 }
 
 // NewServer creates a new API server from the provided modules.
@@ -56,8 +62,12 @@ func NewServer(APIaddr string, requiredUserAgent string, cs modules.ConsensusSet
 	return srv, nil
 }
 
-// Serve listens for and handles API calls. It a blocking function.
+// Serve listens for and handles API calls. It is a blocking function.
 func (srv *Server) Serve() error {
+	// Block the Close() method until Serve() has finished.
+	srv.wg.Add(1)
+	defer srv.wg.Done()
+
 	// stop the server if a kill signal is caught
 	sigChan := make(chan os.Signal)
 	signal.Notify(sigChan, os.Interrupt, os.Kill)
@@ -68,17 +78,31 @@ func (srv *Server) Serve() error {
 		srv.listener.Close()
 	}()
 
-	var errStrs []string
-
 	// The server will run until an error is encountered or the listener is
 	// closed, via either the Close method or the signal handling above.
 	// Closing the listener will result in the benign error handled below.
 	err := srv.apiServer.Serve(srv.listener)
 	if err != nil && !strings.HasSuffix(err.Error(), "use of closed network connection") {
-		errStrs = append(errStrs, fmt.Sprintf("serve err: %v", err))
+		return err
+	}
+	return nil
+}
+
+// Close closes the Server's listener, causing the HTTP server to shut down.
+func (srv *Server) Close() error {
+	var errStrs []string
+
+	// Close the listener, which will cause Server.Serve() to return.
+	if err := srv.listener.Close(); err != nil {
+		errStrs = append(errStrs, fmt.Sprintf("listener err: %v", err))
 	}
 
-	// safely close each module
+	// Wait for Server.Serve() to exit. We wait so that it's guaranteed that the
+	// server has completely closed after Close() returns. This is particularly
+	// useful during testing so that we don't exit a test before Serve() finishes.
+	srv.wg.Wait()
+
+	// Safely close each module.
 	if srv.host != nil {
 		if err := srv.host.Close(); err != nil {
 			errStrs = append(errStrs, fmt.Sprintf("host err: %v", err))
@@ -115,9 +139,4 @@ func (srv *Server) Serve() error {
 		return errors.New(strings.Join(errStrs, "\n"))
 	}
 	return nil
-}
-
-// Close closes the Server's listener, causing the HTTP server to shut down.
-func (srv *Server) Close() error {
-	return srv.listener.Close()
 }

--- a/api/wallet_test.go
+++ b/api/wallet_test.go
@@ -54,10 +54,15 @@ func TestIntegrationWalletGETEncrypted(t *testing.T) {
 
 		server: srv,
 	}
+	errChan := make(chan error)
 	go func() {
 		listenErr := srv.Serve()
-		if listenErr != nil {
-			t.Fatal("API server quit:", listenErr)
+		errChan <- listenErr
+	}()
+	defer func() {
+		err := <-errChan
+		if err != nil {
+			t.Fatalf("API server quit: %v", err)
 		}
 	}()
 	defer st.server.Close()

--- a/api/wallet_test.go
+++ b/api/wallet_test.go
@@ -68,7 +68,7 @@ func TestIntegrationWalletGETEncrypted(t *testing.T) {
 		t.Fatal(err)
 	}
 	if wg.Encrypted {
-		t.Error("Wallet has never been unlocked")
+		t.Error("Wallet has never been encrypted")
 	}
 	if wg.Unlocked {
 		t.Error("Wallet has never been unlocked")
@@ -240,7 +240,7 @@ func TestIntegrationWalletTransactionGETid(t *testing.T) {
 	if testing.Short() {
 		t.SkipNow()
 	}
-	st, err := createServerTester("TestIntegrationWalletGETSiacoins")
+	st, err := createServerTester("TestIntegrationWalletTransactionGETid")
 	if err != nil {
 		t.Fatal(err)
 	}

--- a/build/errors.go
+++ b/build/errors.go
@@ -1,0 +1,23 @@
+package build
+
+import (
+	"errors"
+	"strings"
+)
+
+// JoinErrors concatenates the elements of errs to create a single error. The
+// separator string sep is placed between elements in the resulting error. Nil
+// errors are skipped. If errs is empty or only contains nil elements,
+// JoinErrors returns nil.
+func JoinErrors(errs []error, sep string) error {
+	var strs []string
+	for _, err := range errs {
+		if err != nil {
+			strs = append(strs, err.Error())
+		}
+	}
+	if len(strs) > 0 {
+		return errors.New(strings.Join(strs, sep))
+	}
+	return nil
+}

--- a/build/errors_test.go
+++ b/build/errors_test.go
@@ -1,0 +1,64 @@
+package build
+
+import (
+	"errors"
+	"testing"
+)
+
+// TestJoinErrors tests that JoinErrors only returns non-nil when there are
+// non-nil elements in errs. And tests that the returned error's string the
+// concatenation of all the strings of the elements in errs, in order and
+// separated by sep.
+func TestJoinErrors(t *testing.T) {
+	tests := []struct {
+		errs       []error
+		sep        string
+		wantNil    bool
+		errStrWant string
+	}{
+		// Test that JoinErrors returns nil when errs is nil.
+		{
+			wantNil: true,
+		},
+		// Test that JoinErrors returns nil when errs is an empty slice.
+		{
+			errs:    []error{},
+			wantNil: true,
+		},
+		// Test that JoinErrors returns nil when errs has only nil elements.
+		{
+			errs:    []error{nil},
+			wantNil: true,
+		},
+		{
+			errs:    []error{nil, nil, nil},
+			wantNil: true,
+		},
+		// Test that JoinErrors returns non-nil with the expected string when errs has only one non-nil element.
+		{
+			errs:       []error{errors.New("foo")},
+			sep:        ";",
+			errStrWant: "foo",
+		},
+		// Test that JoinErrors returns non-nil with the expected string when errs has multiple non-nil elements.
+		{
+			errs:       []error{errors.New("foo"), errors.New("bar"), errors.New("baz")},
+			sep:        ";",
+			errStrWant: "foo;bar;baz",
+		},
+		// Test that nil errors are ignored.
+		{
+			errs:       []error{nil, errors.New("foo"), nil, nil, nil, errors.New("bar"), errors.New("baz"), nil, nil, nil},
+			sep:        ";",
+			errStrWant: "foo;bar;baz",
+		},
+	}
+	for _, tt := range tests {
+		err := JoinErrors(tt.errs, tt.sep)
+		if tt.wantNil && err != nil {
+			t.Errorf("expected nil error, got '%v'", err)
+		} else if err != nil && err.Error() != tt.errStrWant {
+			t.Errorf("expected '%v', got '%v'", tt.errStrWant, err)
+		}
+	}
+}

--- a/modules/gateway/gateway.go
+++ b/modules/gateway/gateway.go
@@ -6,7 +6,6 @@ import (
 	"net"
 	"os"
 	"path/filepath"
-	"strings"
 	"time"
 
 	"github.com/NebulousLabs/Sia/build"
@@ -55,12 +54,12 @@ func (g *Gateway) Address() modules.NetAddress {
 
 // Close saves the state of the Gateway and stops its listener process.
 func (g *Gateway) Close() error {
-	var errStrs []string
+	var errs []error
 
-	id := g.mu.RLock()
 	// save the latest gateway state
+	id := g.mu.RLock()
 	if err := g.save(); err != nil {
-		errStrs = append(errStrs, fmt.Sprintf("save error: %v", err))
+		errs = append(errs, fmt.Errorf("save failed: %v", err))
 	}
 	g.mu.RUnlock(id)
 	// send close signal
@@ -69,12 +68,12 @@ func (g *Gateway) Close() error {
 	g.clearPort(g.myAddr.Port())
 	// shut down the listener
 	if err := g.listener.Close(); err != nil {
-		errStrs = append(errStrs, fmt.Sprintf("listener error: %v", err))
+		errs = append(errs, fmt.Errorf("listener.Close failed: %v", err))
 	}
 	// Disconnect from peers.
 	for _, p := range g.Peers() {
 		if err := g.Disconnect(p.NetAddress); err != nil {
-			errStrs = append(errStrs, fmt.Sprintf("peer error: %v", err))
+			errs = append(errs, fmt.Errorf("Disconnect failed: %v", err))
 		}
 	}
 	// Sleep to give time for all goroutines to exit. This is necessary because
@@ -85,13 +84,10 @@ func (g *Gateway) Close() error {
 	// Close the logger. The logger should be the last thing to shut down so that
 	// all other objects have access to logging while closing.
 	if err := g.log.Close(); err != nil {
-		errStrs = append(errStrs, fmt.Sprintf("log error: %v", err))
+		errs = append(errs, fmt.Errorf("log.Close failed: %v", err))
 	}
 
-	if len(errStrs) > 0 {
-		return errors.New(strings.Join(errStrs, "; "))
-	}
-	return nil
+	return build.JoinErrors(errs, "; ")
 }
 
 // New returns an initialized Gateway.

--- a/modules/gateway/gateway.go
+++ b/modules/gateway/gateway.go
@@ -7,6 +7,7 @@ import (
 	"os"
 	"path/filepath"
 	"strings"
+	"time"
 
 	"github.com/NebulousLabs/Sia/build"
 	"github.com/NebulousLabs/Sia/modules"
@@ -70,7 +71,19 @@ func (g *Gateway) Close() error {
 	if err := g.listener.Close(); err != nil {
 		errStrs = append(errStrs, fmt.Sprintf("listener error: %v", err))
 	}
-	// Close the logger.
+	// Disconnect from peers.
+	for _, p := range g.Peers() {
+		if err := g.Disconnect(p.NetAddress); err != nil {
+			errStrs = append(errStrs, fmt.Sprintf("peer error: %v", err))
+		}
+	}
+	// Sleep to give time for all goroutines to exit. This is necessary because
+	// some goroutines write to the logger so we must give them time to exit
+	// before closing the logger.
+	// TODO: block until goroutines exit instead of sleeping.
+	time.Sleep(100 * time.Millisecond)
+	// Close the logger. The logger should be the last thing to shut down so that
+	// all other objects have access to logging while closing.
 	if err := g.log.Close(); err != nil {
 		errStrs = append(errStrs, fmt.Sprintf("log error: %v", err))
 	}

--- a/modules/gateway/gateway.go
+++ b/modules/gateway/gateway.go
@@ -2,12 +2,15 @@ package gateway
 
 import (
 	"errors"
-	"log"
+	"fmt"
 	"net"
 	"os"
+	"path/filepath"
+	"strings"
 
 	"github.com/NebulousLabs/Sia/build"
 	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/persist"
 	"github.com/NebulousLabs/Sia/sync"
 )
 
@@ -38,7 +41,7 @@ type Gateway struct {
 
 	persistDir string
 
-	log *log.Logger
+	log *persist.Logger
 	mu  *sync.RWMutex
 }
 
@@ -51,10 +54,12 @@ func (g *Gateway) Address() modules.NetAddress {
 
 // Close saves the state of the Gateway and stops its listener process.
 func (g *Gateway) Close() error {
+	var errStrs []string
+
 	id := g.mu.RLock()
 	// save the latest gateway state
 	if err := g.save(); err != nil {
-		return err
+		errStrs = append(errStrs, fmt.Sprintf("save error: %v", err))
 	}
 	g.mu.RUnlock(id)
 	// send close signal
@@ -62,19 +67,24 @@ func (g *Gateway) Close() error {
 	// clear the port mapping (no effect if UPnP not supported)
 	g.clearPort(g.myAddr.Port())
 	// shut down the listener
-	return g.listener.Close()
+	if err := g.listener.Close(); err != nil {
+		errStrs = append(errStrs, fmt.Sprintf("listener error: %v", err))
+	}
+	// Close the logger.
+	if err := g.log.Close(); err != nil {
+		errStrs = append(errStrs, fmt.Sprintf("log error: %v", err))
+	}
+
+	if len(errStrs) > 0 {
+		return errors.New(strings.Join(errStrs, "; "))
+	}
+	return nil
 }
 
 // New returns an initialized Gateway.
 func New(addr string, persistDir string) (g *Gateway, err error) {
 	// Create the directory if it doesn't exist.
 	err = os.MkdirAll(persistDir, 0700)
-	if err != nil {
-		return
-	}
-
-	// Create the logger.
-	logger, err := makeLogger(persistDir)
 	if err != nil {
 		return
 	}
@@ -87,7 +97,12 @@ func New(addr string, persistDir string) (g *Gateway, err error) {
 		closeChan:  make(chan struct{}),
 		persistDir: persistDir,
 		mu:         sync.New(modules.SafeMutexDelay, 2),
-		log:        logger,
+	}
+
+	// Create the logger.
+	g.log, err = persist.NewLogger(filepath.Join(g.persistDir, logFile))
+	if err != nil {
+		return nil, err
 	}
 
 	// Register RPCs.

--- a/modules/gateway/persist.go
+++ b/modules/gateway/persist.go
@@ -1,12 +1,17 @@
 package gateway
 
 import (
-	"log"
-	"os"
 	"path/filepath"
 
 	"github.com/NebulousLabs/Sia/modules"
 	"github.com/NebulousLabs/Sia/persist"
+)
+
+const (
+	// nodesFile is the name of the file that contains all seen nodes.
+	nodesFile = "nodes.json"
+	// logFile is the name of the log file.
+	logFile = modules.GatewayDir + ".log"
 )
 
 var persistMetadata = persist.Metadata{
@@ -19,12 +24,12 @@ func (g *Gateway) save() error {
 	for node := range g.nodes {
 		nodes = append(nodes, node)
 	}
-	return persist.SaveFile(persistMetadata, nodes, filepath.Join(g.persistDir, "nodes.json"))
+	return persist.SaveFile(persistMetadata, nodes, filepath.Join(g.persistDir, nodesFile))
 }
 
 func (g *Gateway) load() error {
 	var nodes []modules.NetAddress
-	err := persist.LoadFile(persistMetadata, &nodes, filepath.Join(g.persistDir, "nodes.json"))
+	err := persist.LoadFile(persistMetadata, &nodes, filepath.Join(g.persistDir, nodesFile))
 	if err != nil {
 		return err
 	}
@@ -32,13 +37,4 @@ func (g *Gateway) load() error {
 		g.addNode(node)
 	}
 	return nil
-}
-
-func makeLogger(persistDir string) (*log.Logger, error) {
-	// if the log file already exists, append to it
-	logFile, err := os.OpenFile(filepath.Join(persistDir, "gateway.log"), os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0660)
-	if err != nil {
-		return nil, err
-	}
-	return log.New(logFile, "", log.Ldate|log.Ltime|log.Lmicroseconds|log.Lshortfile), nil
 }

--- a/modules/host/persist.go
+++ b/modules/host/persist.go
@@ -12,9 +12,10 @@ import (
 )
 
 const (
-	// logFile establishes the name of the file that gets used for logging.
+	// settingsFile is the name of the file that stores host settings.
 	settingsFile = "settings.json"
-	logFile      = modules.HostDir + ".log"
+	// logFile establishes the name of the file that gets used for logging.
+	logFile = modules.HostDir + ".log"
 )
 
 // persistMetadata is the header that gets written to the persist file, and is

--- a/modules/miner.go
+++ b/modules/miner.go
@@ -1,6 +1,8 @@
 package modules
 
 import (
+	"io"
+
 	"github.com/NebulousLabs/Sia/types"
 )
 
@@ -71,4 +73,5 @@ type TestMiner interface {
 type Miner interface {
 	BlockManager
 	CPUMiner
+	io.Closer
 }

--- a/modules/miner/miner.go
+++ b/modules/miner/miner.go
@@ -2,6 +2,8 @@ package miner
 
 import (
 	"errors"
+	"fmt"
+	"strings"
 	"sync"
 	"time"
 
@@ -192,16 +194,27 @@ func New(cs modules.ConsensusSet, tpool modules.TransactionPool, w modules.Walle
 // Close terminates all ongoing processes involving the miner, enabling garbage
 // collection.
 func (m *Miner) Close() error {
+	var errStrs []string
+
 	// Save the latest miner state.
 	m.mu.RLock()
 	err := m.save()
 	m.mu.RUnlock()
 	if err != nil {
-		return err
+		errStrs = append(errStrs, fmt.Sprintf("save error: %v", err))
 	}
 
 	m.cs.Unsubscribe(m)
-	return m.log.Close()
+
+	err = m.log.Close()
+	if err != nil {
+		errStrs = append(errStrs, fmt.Sprintf("logger error: %v", err))
+	}
+
+	if len(errStrs) > 0 {
+		return errors.New(strings.Join(errStrs, "; "))
+	}
+	return nil
 }
 
 // checkAddress checks that the miner has an address, fetching an address from

--- a/modules/renter.go
+++ b/modules/renter.go
@@ -7,7 +7,7 @@ import (
 	"github.com/NebulousLabs/Sia/types"
 )
 
-var (
+const (
 	// RenterDir is the name of the directory that is used to store the
 	// renter's persistent data.
 	RenterDir = "renter"

--- a/modules/renter/persist.go
+++ b/modules/renter/persist.go
@@ -6,13 +6,13 @@ import (
 	"encoding/base64"
 	"errors"
 	"io"
-	"log"
 	"os"
 	"path/filepath"
 	"strconv"
 
 	"github.com/NebulousLabs/Sia/build"
 	"github.com/NebulousLabs/Sia/encoding"
+	"github.com/NebulousLabs/Sia/modules"
 	"github.com/NebulousLabs/Sia/persist"
 	"github.com/NebulousLabs/Sia/types"
 )
@@ -20,6 +20,7 @@ import (
 const (
 	PersistFilename = "renter.json"
 	ShareExtension  = ".sia"
+	logFile         = modules.RenterDir + ".log"
 )
 
 var (
@@ -403,12 +404,10 @@ func (r *Renter) initPersist() error {
 	}
 
 	// Initialize the logger.
-	logFile, err := os.OpenFile(filepath.Join(r.persistDir, "renter.log"), os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0660)
+	r.log, err = persist.NewLogger(filepath.Join(r.persistDir, logFile))
 	if err != nil {
 		return err
 	}
-	r.log = log.New(logFile, "", log.Ldate|log.Ltime|log.Lmicroseconds|log.Lshortfile)
-	r.log.Println("STARTUP: Renter has started logging")
 
 	// Load the prior persistance structures.
 	err = r.load()

--- a/modules/renter/renter.go
+++ b/modules/renter/renter.go
@@ -1,10 +1,9 @@
 package renter
 
 import (
-	"log"
-
 	"github.com/NebulousLabs/Sia/modules"
 	"github.com/NebulousLabs/Sia/modules/renter/hostdb"
+	"github.com/NebulousLabs/Sia/persist"
 	"github.com/NebulousLabs/Sia/sync"
 	"github.com/NebulousLabs/Sia/types"
 )
@@ -56,7 +55,7 @@ type Renter struct {
 
 	// resources
 	hostDB hostDB
-	log    *log.Logger
+	log    *persist.Logger
 
 	// variables
 	files         map[string]*file

--- a/modules/wallet/persist.go
+++ b/modules/wallet/persist.go
@@ -2,7 +2,6 @@ package wallet
 
 import (
 	"crypto/rand"
-	"log"
 	"os"
 	"path/filepath"
 
@@ -77,19 +76,6 @@ func (w *Wallet) saveSettings() error {
 	return persist.SaveFile(settingsMetadata, w.persist, filepath.Join(w.persistDir, settingsFile))
 }
 
-// initLog begins logging the wallet, appending to any existing wallet file and
-// writing a startup message to indicate that a new logging instance has been
-// created.
-func (w *Wallet) initLog() error {
-	logFile, err := os.OpenFile(filepath.Join(w.persistDir, logFile), os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0660)
-	if err != nil {
-		return err
-	}
-	w.log = log.New(logFile, "", modules.LogSettings)
-	w.log.Println("STARTUP: Wallet logging has started.")
-	return nil
-}
-
 // initSettings creates the settings object at startup. If a settings file
 // exists, the settings file will be loaded into memory. If the settings file
 // does not exist, a new.persist file will be created.
@@ -122,7 +108,7 @@ func (w *Wallet) initPersist() error {
 	}
 
 	// Start logging.
-	err = w.initLog()
+	w.log, err = persist.NewLogger(filepath.Join(w.persistDir, logFile))
 	if err != nil {
 		return err
 	}

--- a/modules/wallet/wallet.go
+++ b/modules/wallet/wallet.go
@@ -2,12 +2,12 @@ package wallet
 
 import (
 	"errors"
-	"log"
 	"sort"
 	"sync"
 
 	"github.com/NebulousLabs/Sia/crypto"
 	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/persist"
 	"github.com/NebulousLabs/Sia/types"
 )
 
@@ -93,7 +93,7 @@ type Wallet struct {
 	historicClaimStarts map[types.SiafundOutputID]types.Currency
 
 	persistDir string
-	log        *log.Logger
+	log        *persist.Logger
 	mu         sync.RWMutex
 }
 

--- a/persist/log.go
+++ b/persist/log.go
@@ -4,6 +4,7 @@ import (
 	"fmt"
 	"log"
 	"os"
+	"sync"
 
 	"github.com/NebulousLabs/Sia/build"
 )
@@ -14,10 +15,13 @@ import (
 type closeableFile struct {
 	*os.File
 	closed bool
+	mu     sync.RWMutex
 }
 
 // Close closes the file and sets the closed flag.
 func (cf *closeableFile) Close() error {
+	cf.mu.Lock()
+	defer cf.mu.Unlock()
 	// Sanity check - close should not have been called yet.
 	if cf.closed {
 		build.Critical("cannot close the file; already closed")
@@ -32,6 +36,8 @@ func (cf *closeableFile) Close() error {
 
 // Write takes the input data and writes it to the file.
 func (cf *closeableFile) Write(b []byte) (int, error) {
+	cf.mu.RLock()
+	defer cf.mu.RUnlock()
 	// Sanity check - close should not have been called yet.
 	if cf.closed {
 		build.Critical("cannot write to the file after it has been closed")

